File: src/lib/audio/wav.ts
"use client";

/**
 * Whatever the browser recorded, turned into the one format a server can read.
 *
 * `MediaRecorder` does not produce the same thing twice across platforms. Chrome
 * and the WebKitGTK webview give `audio/webm;codecs=opus`; Safari gives
 * `audio/mp4`; the container and the codec inside it are whatever that engine
 * happened to compile in. Posting that straight up means the server has to
 * accept a moving target, and the day somebody's Safari sends AAC to a model
 * that reads Opus, the failure is a transcript that quietly does not arrive.
 *
 * So the shape is decided here, on the client, where the audio already is:
 * 16 kHz mono 16-bit PCM in a WAV container. Uncompressed, ancient, and
 * understood by everything. 16 kHz because speech recognition is trained at it
 * — the band above 8 kHz carries almost no phonetic information, and sending
 * 48 kHz is three times the bytes for nothing.
 *
 * The cost is size: WAV at this rate is 32 KB per second, so a thirty-second
 * slice is about a megabyte. That is why `lib/speech/server.ts` sends slices
 * rather than a whole meeting, and why the route's size limit is what it is.
 *
 * `AudioContext` is present in the desktop webview — measured, see
 * docs/desktop.md — which is what makes this usable there. `SpeechRecognition`
 * is not, which is why it has to be.
 */

/** 16 kHz mono: what a recogniser wants, and a third of the bytes of 48. */
export const WAV_HZ = 16_000;

export interface Encoded {
  /** 16 kHz mono 16-bit PCM, WAV container. */
  wav: Blob;
  /** How long it actually is, from the sample count rather than a clock. */
  seconds: number;
  /**
   * The loudest sample in it, 0..1.
   *
   * Carried alongside the audio because the server needs it. A model handed
   * silence will not say "I heard nothing" — it will write a plausible meeting,
   * which is the failure this whole feature has to be built against. A peak
   * measured here, from the samples themselves, is the one fact about the audio
   * that the model cannot talk its way around. See `/api/listen`.
   */
  peak: number;
}

/**
 * Decode a recording and re-encode it.
 *
 * Throws rather than returning something empty: a caller that cannot encode
 * has nothing to send, and a zero-length WAV posted to a model is the silent
 * slice the route is written to distrust.
 */
export async function toWav(recorded: Blob, hz: number = WAV_HZ): Promise<Encoded> {
  const bytes = await recorded.arrayBuffer();
  if (bytes.byteLength === 0) throw new Error("Nothing was recorded.");

  // Constructed per call and closed at the end. An AudioContext is a hardware
  // handle on most platforms, and one left open per slice is a recorder that
  // runs out of them halfway through a long meeting.
  const context = new AudioContext();
  let decoded: AudioBuffer;
  try {
    decoded = await context.decodeAudioData(bytes);
  } finally {
    void context.close();
  }

  const mono = toMono(decoded);
  const samples = resample(mono, decoded.sampleRate, hz);

  let peak = 0;
  for (const sample of samples) {
    const size = Math.abs(sample);
    if (size > peak) peak = size;
  }

  return {
    wav: wrap(samples, hz),
    seconds: samples.length / hz,
    // Clamped because a decoder may hand back samples slightly outside ±1 and
    // a peak of 1.03 would read as a broken measurement rather than a loud one.
    peak: Math.min(1, peak),
  };
}

/**
 * Every channel averaged into one.
 *
 * Not "take the left channel", which is the shorter version of this and is
 * wrong on any recording where the speaker is on the other side — a laptop with
 * two microphones and somebody sitting to the right of it.
 */
function toMono(buffer: AudioBuffer): Float32Array {
  if (buffer.numberOfChannels === 1) return buffer.getChannelData(0);

  const out = new Float32Array(buffer.length);
  for (let channel = 0; channel < buffer.numberOfChannels; channel++) {
    const data = buffer.getChannelData(channel);
    for (let i = 0; i < data.length; i++) out[i] += data[i];
  }
  for (let i = 0; i < out.length; i++) out[i] /= buffer.numberOfChannels;
  return out;
}

/**
 * Linear interpolation, deliberately.
 *
 * A proper windowed-sinc resampler is better and it is not better *here*: the
 * aliasing it avoids lives above 8 kHz, which is exactly the band being thrown
 * away, and a recogniser is not a listener. What matters is that this is
 * twenty lines with no dependency, runs in a webview with no `OfflineAudio`
 * quirks to work around, and cannot fail on a rate it has not seen.
 */
function resample(input: Float32Array, from: number, to: number): Float32Array {
  if (from === to) return input;

  const ratio = from / to;
  const length = Math.max(1, Math.floor(input.length / ratio));
  const out = new Float32Array(length);
  for (let i = 0; i < length; i++) {
    const at = i * ratio;
    const low = Math.floor(at);
    const high = Math.min(low + 1, input.length - 1);
    const part = at - low;
    out[i] = input[low] * (1 - part) + input[high] * part;
  }
  return out;
}

/** 44 bytes of header, then the samples. The format has not changed since 1991. */
function wrap(samples: Float32Array, hz: number): Blob {
  const bytesPerSample = 2;
  const buffer = new ArrayBuffer(44 + samples.length * bytesPerSample);
  const view = new DataView(buffer);

  const text = (at: number, s: string) => {
    for (let i = 0; i < s.length; i++) view.setUint8(at + i, s.charCodeAt(i));
  };

  const dataBytes = samples.length * bytesPerSample;
  text(0, "RIFF");
  view.setUint32(4, 36 + dataBytes, true);
  text(8, "WAVE");
  text(12, "fmt ");
  view.setUint32(16, 16, true); // PCM header length
  view.setUint16(20, 1, true); // PCM, uncompressed
  view.setUint16(22, 1, true); // mono
  view.setUint32(24, hz, true);
  view.setUint32(28, hz * bytesPerSample, true); // bytes per second
  view.setUint16(32, bytesPerSample, true); // block align
  view.setUint16(34, 8 * bytesPerSample, true); // bits per sample
  text(36, "data");
  view.setUint32(40, dataBytes, true);

  for (let i = 0; i < samples.length; i++) {
    // Clamped before scaling, because a sample of 1.2 scaled and truncated
    // wraps to a large negative — one decoded overshoot becoming an audible
    // click, which a recogniser hears as a consonant.
    const clamped = Math.max(-1, Math.min(1, samples[i]));
    view.setInt16(44 + i * bytesPerSample, clamped * 0x7fff, true);
  }

  return new Blob([buffer], { type: "audio/wav" });
}

/** Base64, in chunks, because `apply` on a megabyte-long array overflows the stack. */
export async function toBase64(blob: Blob): Promise<string> {
  const bytes = new Uint8Array(await blob.arrayBuffer());
  let binary = "";
  const CHUNK = 0x8000;
  for (let i = 0; i < bytes.length; i += CHUNK)
    binary += String.fromCharCode(...bytes.subarray(i, i + CHUNK));
  return btoa(binary);
}

File: desktop/src-tauri/src/record/wav.rs
//! 16 kHz mono 16-bit PCM, in a WAV container.
//!
//! The same format `src/lib/audio/wav.ts` produces in the browser, for the
//! same reason: `/api/listen` accepts one shape of audio, and a server that
//! accepts a moving target fails as a transcript that quietly never arrives.
//! Two encoders exist because there are two capture paths — a browser tab and
//! this window — and neither can call the other.
//!
//! Deliberately not a crate. This is a header of forty-four bytes and a loop,
//! it has not changed since 1991, and the release profile is tuned for size.

/// What a recogniser is trained at. Above 8 kHz there is almost no phonetic
/// information, so 48 kHz would be three times the bytes for nothing.
pub const HZ: u32 = 16_000;

/// Average every channel into one.
///
/// Not "take the first channel", which is the shorter version of this and is
/// wrong on any recording where the speaker sits to the right of a laptop with
/// two microphones.
pub fn to_mono(interleaved: &[f32], channels: u16) -> Vec<f32> {
    if channels <= 1 {
        return interleaved.to_vec();
    }
    let channels = channels as usize;
    interleaved
        .chunks(channels)
        .map(|frame| frame.iter().sum::<f32>() / frame.len() as f32)
        .collect()
}

/// Linear interpolation down to `HZ`.
///
/// A windowed-sinc resampler is better and it is not better here: the aliasing
/// it avoids lives above 8 kHz, which is exactly the band being discarded.
/// What matters is that this has no dependency and cannot fail on a device
/// rate it has not seen — and microphones report all sorts, 44100 and 48000
/// most often but 32000 and 16000 too.
pub fn resample(input: &[f32], from: u32) -> Vec<f32> {
    if from == HZ || input.is_empty() {
        return input.to_vec();
    }
    let ratio = from as f64 / HZ as f64;
    let length = ((input.len() as f64) / ratio).floor().max(1.0) as usize;
    (0..length)
        .map(|i| {
            let at = i as f64 * ratio;
            let low = at.floor() as usize;
            let high = (low + 1).min(input.len() - 1);
            let part = (at - low as f64) as f32;
            input[low] * (1.0 - part) + input[high] * part
        })
        .collect()
}

/// The loudest sample, 0..1.
///
/// This number is why the route can refuse to be lied to. A model handed
/// silence does not say "I heard nothing" — it writes a plausible meeting —
/// and this is arithmetic over samples that the model never sees and cannot
/// talk its way around. See the note at the top of `src/app/api/listen/route.ts`.
pub fn peak(samples: &[f32]) -> f32 {
    samples
        .iter()
        .fold(0.0_f32, |loudest, s| loudest.max(s.abs()))
        .min(1.0)
}

/// Forty-four bytes of header, then the samples.
pub fn encode(samples: &[f32]) -> Vec<u8> {
    let data_bytes = samples.len() * 2;
    let mut out = Vec::with_capacity(44 + data_bytes);

    out.extend_from_slice(b"RIFF");
    out.extend_from_slice(&((36 + data_bytes) as u32).to_le_bytes());
    out.extend_from_slice(b"WAVE");
    out.extend_from_slice(b"fmt ");
    out.extend_from_slice(&16u32.to_le_bytes()); // PCM header length
    out.extend_from_slice(&1u16.to_le_bytes()); // uncompressed
    out.extend_from_slice(&1u16.to_le_bytes()); // mono
    out.extend_from_slice(&HZ.to_le_bytes());
    out.extend_from_slice(&(HZ * 2).to_le_bytes()); // bytes per second
    out.extend_from_slice(&2u16.to_le_bytes()); // block align
    out.extend_from_slice(&16u16.to_le_bytes()); // bits per sample
    out.extend_from_slice(b"data");
    out.extend_from_slice(&(data_bytes as u32).to_le_bytes());

    for sample in samples {
        // Clamped before scaling: a decoded overshoot of 1.2 scaled and
        // truncated wraps to a large negative, which is an audible click, and
        // a recogniser hears a click as a consonant.
        let clamped = sample.clamp(-1.0, 1.0);
        out.extend_from_slice(&((clamped * 32767.0) as i16).to_le_bytes());
    }

    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn the_header_says_what_the_body_is() {
        let wav = encode(&[0.0; 100]);
        assert_eq!(&wav[0..4], b"RIFF");
        assert_eq!(&wav[8..12], b"WAVE");
        assert_eq!(&wav[36..40], b"data");
        assert_eq!(wav.len(), 44 + 200);
        // Channels, rate and bit depth, read back out of the bytes rather than
        // out of the constants that wrote them.
        assert_eq!(u16::from_le_bytes([wav[22], wav[23]]), 1);
        assert_eq!(u32::from_le_bytes([wav[24], wav[25], wav[26], wav[27]]), HZ);
        assert_eq!(u16::from_le_bytes([wav[34], wav[35]]), 16);
    }

    #[test]
    fn resampling_lands_on_the_right_length() {
        // One second at 48 kHz is one second at 16 kHz.
        assert_eq!(resample(&vec![0.0; 48_000], 48_000).len(), 16_000);
        assert_eq!(resample(&vec![0.0; 44_100], 44_100).len(), 16_000);
        // A device already at the target rate is left alone, byte for byte.
        assert_eq!(resample(&vec![0.5; 1_000], HZ).len(), 1_000);
    }

    #[test]
    fn both_channels_are_heard() {
        // Somebody on the right of a laptop: silence on the left channel.
        // Taking the first channel would report an empty room.
        let interleaved = vec![0.0, 0.8, 0.0, 0.8];
        let mono = to_mono(&interleaved, 2);
        assert_eq!(mono, vec![0.4, 0.4]);
        assert!(peak(&mono) > 0.3, "a speaker on one side would be dropped");
    }

    #[test]
    fn a_loud_sample_cannot_wrap_to_a_quiet_one() {
        // A decoder handing back 1.4 must not become a large negative.
        let wav = encode(&[1.4]);
        let sample = i16::from_le_bytes([wav[44], wav[45]]);
        assert!(sample > 32_000, "overshoot wrapped: {sample}");
    }

    #[test]
    fn silence_measures_as_silence() {
        assert_eq!(peak(&[0.0; 50]), 0.0);
        assert!(peak(&[0.0, -0.6, 0.2]) > 0.59);
    }
}
